### PR TITLE
Reduce thread context check overhead

### PR DESCRIPTION
**Pull request by ninja-**
ThreadLocal is a **very significant overhead** in critical areas of the code.
So big it easilly gets to "first hot method by self-time" while benchmarking **netty internals relating to the I/O flow**
Usage of ThreadLocal will never be free no matter the implementation, because of the
Thread.currentThread() call.

Modifications:

Reduce thread context checks when possible by storing recycler's stack handle in places
when the caller is either a thread-local subclass or .get() and .recycle() are only allowed to be called only inside the event loop.

Result:

Less overhead. Improvement visible in profiler and benchmarks.

:exclamation:  Not related to #4402 :exclamation: 

PART 1 : 
- [x] PoolArena,PooledByteBufAllocator,PoolThreadCache,Pooled[Unsafe]DirectByteBuf

- [x]  minor: remove casts in recycler's legacy recycle() method
- [x] cache thread-local stacks for mentioned classes for both get() and recycle() whenever possible
(not always possible to skip thread check for recycle() when we have user code. it's optimized in places when we are sure of the thread i.e. always inside the event loop :+1: )
- [x] modify Recycler to suit the need
- [x] deduplicate the stack handles by assigning them to Stack
- [x] maybe cleanup and get rid of per-item close in PoolThreadCache (make it an argument call)  
Critical Bugs:
- [x] store Pooled*Buf local recycler stacks in the proper place (notes in post below)
- [x] fix MemoryRegionCache.newEntry while not doing a thread-local query somehow... (notes in post below)
- [x] mark outbound buffer entries as unsafe when changing event loops

PART 2:

- [x] Add ChannelHandlerContext.locAlloc() and use it in most handlers
- removes yet another ThreadLocal query by replacing with cache
- stored on EventLoop's object
-  allowed only when inside the correct EventLoop
- [x] move to singlethreadexecutor instead of epolleventloop

PART 3:
- [x] replace IovArrayThreadLocal with IovArray on EpollEventLoop's object to remove unneeded ThreadLocal queries for every write

<s>PART 4:
- [make FastThreadLocal wrap around Java's ThreadLocal to keep the performance on par
-  move the InternalThreadLocalMap's backend to Java's ThreadLocal
- the map is still used here and there
</s>

PART 4:
- [x]  AbstractByteBuf.threadLocal() - adds a flag so it's going to use recycleSameThread(). Call the flag on important handlers like BTBDecoder and MTMEncoder 
or maybee ByteBufAllocator#ioBuffer/directBuffer(true) as a nicer api
(introduce an interface called LocalByteBufAllocator and hook with locAlloc() )

 ------
- [x] fix checkstyle

These 4 commits will say goodbye to ThreadLocals occupying hot methods :smiley:

### CODE REVIEW

**Comment by ninja-** [comment]
@Scottmich @normanmaurer - PTAL

**Comment by Scottmitch** [comment]
@ninja- - WIll look soon. Have you signed the CLA http://netty.io/wiki/developer-guide.html?

**Comment by ninja-** [comment]
@Scottmitch I'll get the CLA soon. I see two other places to get the ThreadLocals out. For example ctx.alloc() (which is used like over and over to rotate the i/o buffers) could skip PooledAllocator's threadCache.get() and use one that's per-eventloop. It wouldn't be fully safe with user code so something like localAlloc() and use that whenever possible in handlers.

**Comment by normanmaurer** [comment]
@ninja- maybe not spent too much time on this before we validated it ? Not want to waste your time if at the end there are issues with the approach. I hope to have time to review tomorrow.

**Comment by ninja-** [comment]
@normanmaurer it's 100% thread-safe. But yeah you should review when you can

**Comment by normanmaurer** [comment]
@ninja- btw if you submit PR's please ensure that these build without errors. This PER has check-style errors.

**Comment by ninja-** [comment]
@normanmaurer yeah as usual...are there any tools to do this at least?

**Comment by normanmaurer** [comment]
mvn clean package should fail if there is a checkstyle error

**Comment by ninja-** [comment]
no I mean like get the formatting fixed with tool. Intellij will break it by itself as you can see with the loops

**Comment by normanmaurer** [comment]
Here is a code style for intellij 

http://netty.io/wiki/setting-up-development-environment.html

**Comment by ninja-** [comment]
Ok I'll fix the checkstyle.
I have just finished the second part which creates ctx.locAlloc() and uses that in most handlers; this should drop even more ThreadLocal overhead. That will be in a seperate commit.

**Comment by ninja-** [comment]
https://github.com/netty/netty/pull/4417
> I was looking forward to doing some deep digging, but got disappointed :(.

No worries @buchgr @normanmaurer, it still causes noticable overhead that I am going to cover (✌ ﾟ ∀ ﾟ)☞

**Comment by ninja-** [comment]
I don't think the http test could be related to the changes

**Comment by ninja-** [comment]
Found some issues to fix already, additional reviews welcome

File: README.md [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Netty Project
+# Netty Project blabla
```
Comment (normanmaurer) [inline README.md]: ?

Comment (ninja-) [inline README.md]: ignore, I did an empty commit to force the bot to re-run the tests

**Comment by normanmaurer** [comment]
@ninja- so just to ensure I get the idea... Basically you want to remove the "ThreadLocal" lookup when you know that for example everything is always done from within the EventLoop of a Channel. Like for example in ChannelOutboundBuffer. Is this correct ?

I did not have the time yet to have a deep look but one thing that I don't like from a quick look is that you "pulled" the ByteBufAllocator into the EventLoop. An EventLoop should not be aware an ByteBufAllocator. This blur the separation of different parts of the API and different responsibilities.

**Comment by ninja-** [comment]
> I did not have the time yet to have a deep look but one thing that I don't like from a quick look is that you "pulled" the ByteBufAllocator into the EventLoop. An EventLoop should not be aware an ByteBufAllocator. This blur the separation of different parts of the API and different responsibilities.

Hmm I did this to reduce memory usage. Like, the local allocator instance needs to be "per something". I didn't want to make an instance per ChannelPipeline etc. so if you have an idea on how to hide it from EventLoop then I could get it out of EventLoop in patch #2

**Comment by ninja-** [comment]
@Scottmitch @normanmaurer fyi finished my list

**Comment by ninja-** [comment]
@normanmaurer wdyt?

**Comment by Scottmitch** [comment]
@ninja- - Can you report benchmark results which demonstrate the improvements claimed by this PR?

**Comment by ninja-** [comment]
@Scottmich you know this kind of improvement that benchmark doesn't show but profiler does? That's one of them...

**Comment by normanmaurer** [comment]
@ninja- sorry if this sounds harsh but without a proof I can not accept this. This PR touches so many places that I'm not feeling "convinced" enough without seeing it myself. Maybe it would be easier if you just do "small / isolated" changes, proof it and open a PR.

**Comment by ninja-** [comment]
hello @normanmaurer any chance you could revisit this PR? 
(it's been working correctly for me ever since - I am still using it)

I remember you tried to pull the 3rd patch once and quickly reverted when figuring out it's not thread safe in ForkJoinPool. Well I really can't agree on that. Could you explain that?

**Pull request closed without merging**